File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.cpp
class Solution {
public:
int numSubarraysWithSum(vector<int> &nums, int goal)
{
    map<long long, int> mp;
    // if (goal != 0)
    mp[0] = 1;
    int ans = 0;
    long long sum = 0;
    for (int i = 0; i < nums.size(); i++)
    {
        sum += nums[i];
        long long diff = sum - goal;
        if (mp.count(diff))
        {
            ans += mp[diff];
        }
        mp[sum]++;
    }
    return ans;
}
};

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int k) {
        int count = 0;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int total = 0;
        mp.put(0, 1);
        for (int i : nums) {
            total += i;
            count += mp.getOrDefault(total - k, 0);
            mp.put(total, mp.getOrDefault(total, 0) + 1);
        }
        return count;
    }
}